### PR TITLE
ETL DB: add judge_case_reviews and decision_documents tables

### PR DESCRIPTION
**Pull request by yoomlam**
Connects #16592

This PR adds new tables in the ETL DB for Paul S.'s D.O.C. (Decision Output Calculator = "required decision output of production attorneys"). 

### Description
New ETL DB tables: `judge_case_reviews` and `decision_documents`
and syncers to populate them.

Bonus: update comments on Caseflow DB's `decision_documents` table.

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Paul is able to query data for D.O.C.

### Testing Plan
```ruby
make etl-migrate
make etl-test-prepare
bundle exec rspec spec/services/etl/decision_document_syncer_spec.rb
bundle exec rspec spec/services/etl/judge_case_review_syncer_spec.rb
```

### Deployment Plan
1. Once deployed, manually test ETL job and monitor for errors
2. Check new ETL DB tables and test Paul's (new) queries against them
3. Have Paul run queries against the Redshift copy of the ETL DB  and test his D.O.C.

### Database Changes

* [x] Add typical timestamps (`created_at`, `updated_at`) for new tables
* [x] Update column comments; include a "PII" prefix to indicate definite or potential [PII data content](https://github.com/department-of-veterans-affairs/appeals-team/blob/master/caseflow-team/0-how-we-work/pii-handbook.md#what-is-pii)
* [x] Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`)
* [x] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [x] Perform query profiling (eyeball Rails log, check bullet and fasterer output)
* [x] Add appropriate indexes (especially for foreign keys, polymorphic columns, unique constraints, and Rails scopes)
* [ ] ~Run `make check-fks`; add any missing foreign keys or add to `config/initializers/immigrant.rb`~
* [ ] ~Run `make docs` (after running `make migrate`) to update DB schema docs~
* [x] Post this PR in #appeals-schema with a summary
* [ ] ~Document any non-obvious semantics or logic useful for interpreting database data at [Caseflow Data Model and Dictionary](https://github.com/department-of-veterans-affairs/caseflow/wiki/Caseflow-Data-Model-and-Dictionary)~

### CODE REVIEW

**Comment by yoomlam** [comment]
To facilitate Paul S.'s D.O.C. (Decision Output Calculator = "required decision output of production attorneys"), two new ODS tables where added: `judge_case_reviews` and `decision_documents`.

See descriptions for the table columns here:
* `judge_case_reviews`: https://github.com/department-of-veterans-affairs/caseflow/blob/e539627723508f40acb10ac005ddf5fedc6aa87d/db/etl/schema.rb#L364-L390
* `decision_documents`: https://github.com/department-of-veterans-affairs/caseflow/blob/e539627723508f40acb10ac005ddf5fedc6aa87d/db/etl/schema.rb#L122-L143

(Hint: the columns are organized (instead of alphabetized) in the PR: [`judge_case_reviews`](https://github.com/department-of-veterans-affairs/caseflow/pull/16617/files#diff-b62610a5ff7e2fd90779b076cd4096788b6c4b84bfaf58ffeabe9b3bdf998d1bR5), [`decision_documents`](https://github.com/department-of-veterans-affairs/caseflow/pull/16617/files#diff-71ff3d4faf47fd47806061ac89027a1b6dfbdb89809050f08ef4757fbeb5eccdR5))

File: app/models/appeal.rb [hunk 1]
```diff
@@ -214,8 +214,12 @@ def latest_attorney_case_review
       .order(:created_at).last
   end
 
+  def latest_judge_case_review
+    @latest_judge_case_review ||= JudgeCaseReview.where(task_id: tasks.pluck(:id)).order(:created_at).last
+  end
+
   def reviewing_judge_name
-    task = tasks.not_cancelled.of_type(:JudgeDecisionReviewTask).order(created_at: :desc).first
+    task = tasks.not_cancelled.of_type(:JudgeDecisionReviewTask).order(:created_at).last
```
Comment (lowellrex) [inline app/models/appeal.rb:222]: Smart!

**Review by lowellrex — APPROVED** [review]
LGTM! Thanks Yoom!

**Comment by BVA-IDT** [comment]
@yoomlam I just ran the research tools query against the old query for the month of August.  I noticed that several of the attorney/judge fields are blank in the new (ETL) query.   (Does not happen in the old query).  Also noted 8098 results with old query, 7322 results for new query.  I can give you the files when old vs new is run if you want.

Also did you incorporate the value transformation for the outcomes and categories?  Those were in the sed command I gave.

Redshift new query (added date limitation and replaced invalid CONCAT)
`SELECT
  docs.citation_number AS "Citation nr.",
  docs.docket_number AS "Docket nr.",
  to_char(docs.decision_date, 'mm/dd/yy') AS "Decision Date",
  judge.sattyid || ' ' || judge.full_name AS "VLJ number and name",
  atty.sattyid || ' ' || atty.full_name AS "Attorney number and name",
  issue.disposition AS Disposition,
  issue.benefit_type AS "Program area desc",
  REPLACE(REPLACE(issue.description, CHR(13), ' '), CHR(10),'') AS "Issue code desc",
  REPLACE(REPLACE(issue.diagnostic_code, CHR(13), ' '), CHR(10), '') AS "Issue lev1 desc",
  '' AS "Issue lev2 desc",
  '' AS "Issue lev3 desc"
FROM ods.decision_documents AS docs
LEFT JOIN ods.decision_issues AS issue
  ON docs.appeal_id = issue.decision_review_id
  AND docs.appeal_type = issue.decision_review_type
LEFT JOIN ods.users AS judge
  ON judge.id = docs.judge_user_id
LEFT JOIN ods.users AS atty
  ON atty.id = docs.attorney_user_id
WHERE docs.decision_date >= '08/01/2021'::DATE and docs.decision_date < '09/01/2021'::DATE;`

**Comment by yoomlam** [comment]
Quick response:
>  I can give you the files when old vs new is run if you want.

Yes, please send to my VA email.

> Also did you incorporate the value transformation for the outcomes and categories? Those were in the sed command I gave.

No, I did not. I didn't want to transform it too much from the original data to avoid confusion in case the ETL table is used for other purposes. What I can do is create new columns that applies the `sed` commands.

**Comment by BVA-IDT** [comment]
For the sed commands what I'd really like is a view that I could just do a `SELECT * FROM TABLE WHERE DATE<=....`  Doesn't have to be a real table, but something that had the query and transformation baked it.

**Comment by BVA-IDT** [comment]
@yoomlam 
## Original Query
```
SELECT tasks.appeal_id, 
MAX(date_trunc('day', review.created_at)) "updated"
, (SELECT COUNT(issues.id) FROM decision_issues as issues WHERE issues.decision_review_type='Appeal' AND issues.disposition IN ('denied','remanded','allowed') and issues.decision_review_id=tasks.appeal_id AND issues.deleted_at IS NULL) "ard_issues"
, (SELECT COUNT(issues.id) FROM decision_issues as issues WHERE issues.decision_review_type='Appeal' AND NOT issues.disposition IN ('denied','remanded','allowed') and issues.decision_review_id=tasks.appeal_id AND issues.deleted_at IS NULL) "other_issues"
, (SELECT attorney.overtime FROM attorney_case_reviews as attorney INNER JOIN tasks AS tasks_two ON attorney.task_id=tasks_two.id WHERE tasks_two.appeal_id=tasks.appeal_id LIMIT 1) "overtime"
FROM tasks INNER JOIN judge_case_reviews AS review ON tasks.id=review.task_id 
WHERE review.attorney_id=(SELECT id from users WHERE css_id='VACOSAINDP' limit 1) AND tasks.status='completed' 
GROUP BY tasks.appeal_id
ORDER BY updated;
```
## Original Results
```
 appeal_id |       updated       | ard_issues | other_issues | overtime
-----------+---------------------+------------+--------------+----------
      5478 | 2019-06-11 00:00:00 |          1 |            0 | false
      5674 | 2019-10-21 00:00:00 |         10 |            0 | false
     14838 | 2019-10-21 00:00:00 |          2 |            0 | false
      7096 | 2019-11-19 00:00:00 |          1 |            0 | false
      7020 | 2019-11-19 00:00:00 |          1 |            0 | false
     19490 | 2019-11-19 00:00:00 |          2 |            0 | false
      6569 | 2019-11-21 00:00:00 |          2 |            0 | false
      8903 | 2019-12-20 00:00:00 |          3 |            0 | false
      6628 | 2019-12-20 00:00:00 |          1 |            0 | false
      1402 | 2020-02-04 00:00:00 |          5 |            0 | false
      4387 | 2020-02-05 00:00:00 |         11 |            0 | false
     12731 | 2020-03-19 00:00:00 |          1 |            0 | false
     10816 | 2020-03-23 00:00:00 |          0 |            2 | false
     36435 | 2020-04-08 00:00:00 |          7 |            0 | false
     50611 | 2020-04-13 00:00:00 |          1 |            0 | false
     15555 | 2020-07-17 00:00:00 |          4 |            0 | false
     78194 | 2020-08-21 00:00:00 |          2 |            0 | false
     12601 | 2020-11-13 00:00:00 |          1 |            0 | false
     43141 | 2020-12-02 00:00:00 |          4 |            0 | false
     72790 | 2021-02-04 00:00:00 |          3 |            0 | false
     83297 | 2021-02-11 00:00:00 |          3 |            0 | false
     79341 | 2021-03-23 00:00:00 |          7 |            0 | false
     50580 | 2021-04-09 00:00:00 |          1 |            0 | false
```
## New Query
```
SELECT
  docs.appeal_id, docs.appeal_type,
  to_char(judge_cr.review_created_at, 'mm/dd/yy') AS "updated",
  atty_cr.overtime AS "Attorney overtime",
  COUNT(CASE WHEN issues.disposition IN ('denied','remanded','allowed') THEN 1 ELSE null END) AS "ard_issues",
  COUNT(CASE WHEN issues.disposition IN ('denied','remanded','allowed') THEN null ELSE 1 END) AS "other_issues"
FROM ods.decision_documents AS docs
LEFT JOIN ods.decision_issues AS issues
  ON docs.appeal_id = issues.decision_review_id
  AND docs.appeal_type = issues.decision_review_type
  AND issues.issue_deleted_at IS NULL
LEFT JOIN ods.attorney_case_reviews AS atty_cr
  ON docs.appeal_id = atty_cr.appeal_id
  AND docs.appeal_type = atty_cr.appeal_type
LEFT JOIN ods.judge_case_reviews AS judge_cr
  ON docs.appeal_id = judge_cr.appeal_id
  AND docs.appeal_type = judge_cr.appeal_type
LEFT JOIN ods.users AS judge
  ON judge.user_id = docs.judge_user_id
LEFT JOIN ods.users AS atty
  ON atty.user_id = docs.attorney_user_id
WHERE
  atty.sattyid = '1826' and docs.appeal_type = 'Appeal'
GROUP BY docs.appeal_id, docs.appeal_type,
  judge_cr.review_created_at,
  atty_cr.overtime
ORDER BY updated;
```
## New Results
```
 appeal_id | appeal_type | updated  | attorney overtime | ard_issues | other_issues
-----------+-------------+----------+-------------------+------------+--------------
      1402 | Appeal      | 02/04/20 | false             |         30 |            0
     72790 | Appeal      | 02/04/21 | false             |          6 |            0
      4387 | Appeal      | 02/05/20 | false             |         22 |            0
     83297 | Appeal      | 02/11/21 | false             |          6 |            0
     12731 | Appeal      | 03/19/20 | false             |          2 |            0
     10816 | Appeal      | 03/23/20 | false             |          0 |            3
     79341 | Appeal      | 03/23/21 | false             |          7 |            0
     36435 | Appeal      | 04/08/20 | false             |         14 |            0
     50580 | Appeal      | 04/09/21 | false             |          1 |            0
     50611 | Appeal      | 04/13/20 | false             |          2 |            0
      5478 | Appeal      | 06/11/19 | false             |          1 |            0
     15555 | Appeal      | 07/17/20 | false             |          4 |            0
     78194 | Appeal      | 08/21/20 | false             |          2 |            0
      5674 | Appeal      | 10/21/19 | false             |         10 |            0
     14838 | Appeal      | 10/21/19 | false             |          2 |            0
     12601 | Appeal      | 11/13/20 | false             |          2 |            0
      7020 | Appeal      | 11/19/19 | false             |          1 |            0
     19490 | Appeal      | 11/19/19 | false             |          2 |            0
      7096 | Appeal      | 11/19/19 | false             |          1 |            0
      6569 | Appeal      | 11/21/19 | false             |          2 |            0
     43141 | Appeal      | 12/02/20 | false             |          8 |            0
      8903 | Appeal      | 12/20/19 | false             |          6 |            0
      6628 | Appeal      | 12/20/19 | false             |          2 |            0
(23 rows)
```

Some agreement, but those issue counts are rather off  in some of the instances.

**Comment by yoomlam** [comment]
"New Query v2" at the bottom of #16772 should now show consistent issue counts.

**Comment by yoomlam** [comment]
> Quick response:
> 
> > I can give you the files when old vs new is run if you want.
> 
> Yes, please send to my VA email.
> 
> > Also did you incorporate the value transformation for the outcomes and categories? Those were in the sed command I gave.
> 
> No, I did not. I didn't want to transform it too much from the original data to avoid confusion in case the ETL table is used for other purposes. What I can do is create new columns that applies the `sed` commands.

Instead of creating additional columns with duplicative data, I came up with this straight-forward SQL query that includes the `sed` transforms:

```sql
SELECT
  docs.citation_number AS "Citation nr.",
  docs.docket_number AS "Docket nr.",
  to_char(docs.decision_date, 'mm/dd/yy') AS "Decision Date",
  judge.sattyid || ' ' || judge.full_name AS "VLJ number and name",
  atty.sattyid || ' ' || atty.full_name AS "Attorney number and name",

  INITCAP(REPLACE(issue.disposition,'_',' ')) AS Disposition,
  CASE 
    WHEN issue.benefit_type='voc_rehab' THEN 'VR&E'
    WHEN issue.benefit_type='vha' THEN 'Medical'
    WHEN issue.benefit_type='nca' THEN 'NCA Burial Benefits (38 CFR Part 38)'
    ELSE INITCAP(issue.benefit_type)
  END AS "Program area desc",
  
  REPLACE(REPLACE(issue.description, CHR(13), ' '), CHR(10),'') AS "Issue code desc",
  REPLACE(REPLACE(issue.diagnostic_code, CHR(13), ' '), CHR(10), '') AS "Issue lev1 desc",
  '' AS "Issue lev2 desc",
  '' AS "Issue lev3 desc"
FROM ods.decision_documents AS docs
LEFT JOIN ods.decision_issues AS issue
  ON docs.appeal_id = issue.decision_review_id
  AND docs.appeal_type = issue.decision_review_type
LEFT JOIN ods.users AS judge
  ON judge.id = docs.judge_user_id
LEFT JOIN ods.users AS atty
  ON atty.id = docs.attorney_user_id
WHERE docs.decision_date >= '08/01/2021'::DATE and docs.decision_date < '09/01/2021'::DATE
```

**Comment by yoomlam** [comment]
@BVA-IDT I believe I've responded to all the issues listed in https://github.com/department-of-veterans-affairs/caseflow/pull/16617#issuecomment-914560127

**Pull request merged**